src/systems/a320_systems/src/flight_warning/runtime/monitor.rs

```diff
@@ -0,0 +1,347 @@
+    //cleared: HashSet<WarningCode>,
+    canceled: HashSet<WarningCode>,
+    c_chord: bool,
+    minimum_generated: bool,
+    hundred_above_generated: bool,
+    auto_call_out_generated: bool,
+    inter_audio: bool,
+
+    remaining_synthetic_voice_files: VecDeque<Vec<VoiceFile>>,
+    requested_synthetic_voice: Option<Vec<VoiceFile>>,
+}
+
+impl A320FWCMonitor {
+    pub(super) fn update(
+        &mut self,
+        delta: Duration,
+        parameters: &impl A320FWCMonitorParameters,
+        synthetic_voice_ready: bool,
+    ) {
+        let requested_warnings = parameters.requested_warnings();
+        let warnings = self.determine_active_warnings(delta, &requested_warnings);
+        // cancel level 3 audio
+        if parameters.mw_cancel_pulse_up() {
+            if warnings.contains(&WarningCode::new(22, 00, 050)) {
+                self.canceled.insert(WarningCode::new(22, 00, 050));
+            }
+        }
+
+        let mut new_cancel = HashSet::new();
+        for canceled in self.canceled.drain() {
+            // only keep if still active
+            if requested_warnings.contains(&canceled) {
+                new_cancel.insert(canceled);
+            }
+        }
+        self.canceled = new_cancel;
+
+        self.update_auto_call_outs(parameters, &warnings, synthetic_voice_ready);
+        self.update_sounds(&warnings);
+    }
+
+    /// This method updates the auto call outs which are presented to the pilots as Synthetic Voice.
+    /// Examples of these call outs are critical warnings (Stall, Speed Windshear) and more
+    /// informational like altitude callouts (Forty, Thirty, Twenty, Retard).
+    /// At every update tick, the monitor will check if the audio subsystem is ready. If it is, it
+    /// can queue up a sequence of callouts.
+    fn update_auto_call_outs(
+        &mut self,
+        parameters: &impl A320FWCMonitorParameters,
+        warnings: &Vec<WarningCode>,
+        synthetic_voice_ready: bool,
+    ) {
+        // reset the callout started pulses
+        self.minimum_generated = false;
+        self.hundred_above_generated = false;
+        self.auto_call_out_generated = false;
+        self.inter_audio = false;
+
+        // can we even start a new sound?
+        if !synthetic_voice_ready || self.requested_synthetic_voice.is_some() {
+            return;
+        }
+
+        if !self.remaining_synthetic_voice_files.is_empty() {
+            self.requested_synthetic_voice = self.remaining_synthetic_voice_files.pop_front();
+            return;
+        }
+
+        // is a voice sequence still in progress?
+        if warnings.contains(&WarningCode::new(34, 00, 360)) {
+            self.auto_call_out_generated = true;
+            self.remaining_synthetic_voice_files = VecDeque::from(vec![
+                vec![VoiceFile::Ten],
+                vec![VoiceFile::Retard],
+                vec![VoiceFile::Pause],
+                vec![VoiceFile::Pause],
+            ]);
```

Comment:
Yeah, it's about inhibitability. The outer vec is a list of callouts ("Ten", "Fifty-one", "Twenty Retard") and the inner vec is all the atoms that make up a single call out(`["Ten"]`, `["Fifty", "One"]`, `["Twenty", "Retard"]`).
While a callout can fully be inhibited, once we've commited to starting it we will always play it out - otherwise you might end up with wrong callouts like "Fifty-Ten".

I you have a suggestions for a more suitable data structure I'd be happy to hear about it!

Reply:
Many thoughts pop up in my head. I'll just write down something. You can judge if it's valuable 😀. 

You could represent the type of audio sequence with an enum:
```rust
pub(super) enum AudioSequence {
    Interruptible(Vec<VoiceFile>),
    Uninterruptible(Vec<VoiceFile>),
}
```
With some `From<AudioSequence> for VecDeque<Vec<VoiceFile>>` implementation.

This would change:
```rust
            self.remaining_synthetic_voice_files = VecDeque::from(vec![
                vec![VoiceFile::Ten],
                vec![VoiceFile::Retard],
                vec![VoiceFile::Pause],
                vec![VoiceFile::Pause],
            ]);
```
into:
```rust
            self.remaining_synthetic_voice_files = AudioSequence::Interruptible(vec![
                VoiceFile::Ten,
                VoiceFile::Retard,
                VoiceFile::Pause,
                VoiceFile::Pause,
            ])
            .into();
```
One can then hide more implementation details and make things a bit nicer with something like:
```rust
impl AudioSequence {
    pub(super) fn interruptible(files: &[VoiceFile]) -> Self {
        Self::Interruptible(files.to_vec())
    }

    pub(super) fn include(mut self, files: &[VoiceFile]) -> Self {
        self.files().extend(files);

        self
    }

    pub(super) fn pause(mut self, times: u8) -> Self {
        for _ in 1..=times {
            self.files().push(VoiceFile::Pause)
        }

        self
    }

    fn files(&mut self) -> &mut Vec<VoiceFile> {
        match self {
            AudioSequence::Interruptible(files) => files,
            AudioSequence::Uninterruptible(files) => files,
        }
    }
}
```
I added the `include` function in the above for situations where you'd want to pause and then include other voice files, if that never happens, then it's probably not needed.

Resulting in:
```rust
            self.remaining_synthetic_voice_files =
                AudioSequence::interruptible(&[VoiceFile::Ten, VoiceFile::Retard])
                    .pause(2)
                    .into();
```
Maybe `self.remaining_synthetic_voice_files` can be a type that wraps `VecDeque<_>` instead of exposing it, such that one could write:
```rust
            self.remaining_synthetic_voice_files +=
                AudioSequence::interruptible(&[VoiceFile::Ten, VoiceFile::Retard]).pause(2);
```
Not sure if an `Add` operator can work on type `Into<VecDeque<Vec<VoiceFile>>>`, but if it can then that would be even more generic than `Add` on `AudioSequence`.

Of course if the order of audio is always voice files followed by 0..n pauses and nothing else, one could also just put this in the constructor function and write:
```rust
            self.remaining_synthetic_voice_files +=
                AudioSequence::interruptible(&[VoiceFile::Ten, VoiceFile::Retard], 2);
```

Reply:
This is roughly how we're doing it now, although interruptible is implemented as a trait. We're also using a new type of `impl SyntheticVoice` for much longer and only resolve to a `Vec` much later.